### PR TITLE
TaggedObject encoding for Nullary constructor does not have "contents" field

### PR DESCRIPTION
**Pull request by devwout**
As requested in #300 , implemented this for both GHC Generics and TemplateHaskell .

### CODE REVIEW

File: Data/Aeson/Types/Generic.hs [hunk 1]
```diff
@@ -201,7 +201,6 @@ instance ( IsRecord      a isRecord
         (builder tagFieldName <>
          B.char7 ':' <>
          builder (constructorTagModifier opts (conName (undefined :: t c a p)))) <>
-        B.char7 ',' <>
```
Comment (bos) [inline Data/Aeson/Types/Generic.hs]: This is a bit mysterious. Can you explain what's going on here?

Comment (bos) [inline Data/Aeson/Types/Generic.hs]: I think I see what's going on: you moved the comma insertion elsewhere. Is that correct?

(Ideally, the movement of comma insertion would have been a separate standalone commit, and it would thus have been obvious what was going on.)

Comment (devwout) [inline Data/Aeson/Types/Generic.hs]: Yes indeed, I moved the comma insertion. Otherwise, the encoding for nullary constructor would be `{"tag": "X",}` (trailing comma). I agree a separate commit would have made this more clear, I'll keep it in mind.

File: Data/Aeson/Types/Generic.hs [hunk 2]
```diff
@@ -211,11 +210,18 @@ class TaggedObject' f isRecord where
     taggedObjectPairs' :: Options -> String -> f a -> Tagged isRecord [Pair]
     taggedObjectEnc' :: Options -> String -> f a -> Tagged isRecord B.Builder
 
+instance OVERLAPPING_ TaggedObject' U1 True where
```
Comment (basvandijk) [inline Data/Aeson/Types/Generic.hs]: Hi @devwout, I think it would be better not to treat U1 as a record. I think the following is simpler.

Encoding:

```haskell
instance OVERLAPPING_ TaggedObject' U1 False where
    taggedObjectPairs' _opts _contentsFieldName =
        Tagged . const []
    {-# INLINE taggedObjectPairs' #-}

    taggedObjectEnc' _opts _contentsFieldName =
        Tagged . const mempty
    {-# INLINE taggedObjectEnc' #-}
```

Decoding:

```haskell
instance OVERLAPPING_ FromTaggedObject'' U1 False where
    parseFromTaggedObject'' _opts _contentsFieldName = Tagged . const (pure U1)
    {-# INLINE parseFromTaggedObject'' #-}
```

Could you try if that works and update your PR?

**Comment by devwout** [comment]
Hi @basvandijk 

I updated my pull request based on your suggestions and it still works as expected. Explicitly specifying the case for nullary constructors is indeed simpler.

If you prefer me to amend these two commits to the previous one, tell me. We are all busy, I understand this took a while.

**Comment by bos** [comment]
@devwout if you could please squash and rebase your commits, I'd appreciate it. Thank you.

**Comment by devwout** [comment]
Hi @bos 

The pull request is rebased and squashed into two commits. I think it is best to commit the tests for existing functionality separately.
